gustaf/ffd.py

```diff
@@ -1,4 +1,571 @@
+            # v_ind = self._original_mesh.select_vertices(
+            #     "xyz_range",
+            #     criteria=utils.bounds_to_ranges(
+            #         self._naive_spline.knot_vectors_bounds
+            #     ),
+            # )
+
+            # if len(v_ind) > 0:
+            #     self._logd(
+            #         "  There are {v}/{vt} vertices to deform.".format(
+            #             v=len(v_ind),
+            #             vt=len(self._mesh.vertices),
+            #         )
+            #     )
+
+            #     # Add new axis if current spline maps to higher dim
+            #     if (
+            #         self._spline.physical_dim != self._mesh.vertices.shape[1]
+            #         and self._spline.physical_dim > self._spline.parametric_dim
+            #     ):
+            #         for _ in range(
+            #             self.spline.physical_dim 
+            #             - self.spline.parametric_dim
+            #         ):
+            #             self._mesh.vertices = np.hstack(
+            #                 (
+            #                     self._original_mesh.vertices,
+            #                     np.zeros((self._mesh.vertices.shape[0], 1)),
+            #                 )
+            #             )
+
+            #     # Here, no need to use q_vertices, since we don't deform
+            #     # original
+            #     self._mesh.vertices[v_ind] = self._spline.evaluate(
+            #         self._original_mesh.vertices[v_ind]
+            #     )
+            #     self._logd("partial FFD successful.")
+            # else:
+            #    self._logw(
+            #         "Spline does not contain any mesh. Nothing to deform."
+            #     )
+
+            return
+
+        self._logd("Applying FFD...")
+        # Here, we take _q_vertices, due to possible scale/offset.
+        self._mesh.vertices = self._spline.evaluate(
+            self._q_vertices
+        )
+        self._logd("FFD successful.")
+        self.calculated = True
+    
+
+    @property
+    def control_points(self):
+        """
+        Returns current spline's control points. Can't and don't use this to
+        directly manipulate control points: it won't update
+
+        Returns
+        --------
+        self._spline.control_points: np.ndarray
+        """
+        return self._spline.control_points
+
+    @control_points.setter
+    def control_points(self,
+                       control_points: Union[List[List[float]], np.ndarray]
+                       ):
+        """
+        Sets control points and deforms mesh.
+
+        Parameters
+        -----------
+        control_points: np.ndarray
+
+        Returns
+        --------
+        None
+        """
+        assert self._spline.control_points.shape == control_points.shape,\
+            "Given control points' shape does not match current ones!"
+
+        self._spline.control_points = control_points.copy()
```

Comment:
Is this copy really necessary?

Reply:
Also, is_calculated should be set to false here imo

Reply:
calculated is removed completely